### PR TITLE
[15.0][MIG] pos_lot selection

### PR DESCRIPTION
**Pull request by i-vyshnevska**
related https://github.com/OCA/pos/pull/827

### CODE REVIEW

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    const Registries = require("point_of_sale.Registries");
+    const {_lt} = require("@web/core/l10n/translation");
+
+    class LotSelectorPopup extends AbstractAwaitablePopup {
+        constructor() {
+            super(...arguments);
+            // If there's a product, get lots available related to this product
+            this.slected_items = [];
+            if (this.props.product) {
+                this.lots = this.env.pos.db.lot_by_product_id[this.props.product.id];
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: Can we please avoid loading lots in the POS session and do and RPC call to get the list of lots from the popup instead?

File: pos_lot_selection/static/src/js/popup/ProductScreen.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+odoo.define("pos_lot_selection.ProductScreen", function (require) {
+    "use strict";
+
+    const ProductScreen = require("point_of_sale.ProductScreen");
+    const Registries = require("point_of_sale.Registries");
+    const NumberBuffer = require("point_of_sale.NumberBuffer");
+
+    const PosLotSaleProductScreen = (ProductScreen) =>
+        class extends ProductScreen {
+            async _clickProduct(event) {
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/ProductScreen.js]: Overriding `_clickProduct` won't provide this module's feature if the product barcode was scanned.. I feel like it would be better to override `_getAddProductOptions` or even `showPopup` instead.

File: pos_lot_selection/static/src/js/popup/ProductScreen.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                const product = event.detail;
+                if (product.tracking !== "none") {
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/ProductScreen.js]: With this condition, we're not calling `_getAddProductOptions` which is an issue if for example you have a product tracked by lot that needs to be weighed with an electronic scale.

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js (hunk 1 above)
Comment (cvinh) [reply pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: AFAIK, pos has been designed to work offline, if you call RPC in real time, wouldn't you break that feature ?

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    const AbstractAwaitablePopup = require("point_of_sale.AbstractAwaitablePopup");
+    const Registries = require("point_of_sale.Registries");
+    const {_lt} = require("@web/core/l10n/translation");
+
+    class LotSelectorPopup extends AbstractAwaitablePopup {
+        constructor() {
+            super(...arguments);
+            // If there's a product, get lots available related to this product
+            this.slected_items = [];
```
Comment (cvinh) [inline pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: Why do you define slect_items ? I don't see that you use that later

File: pos_lot_selection/static/src/js/db.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        init: function () {
+            this._super.apply(this, arguments);
+            this.lot_by_product_id = {};
+        },
+        add_lots: function (lots) {
+            if (!lots instanceof Array) {
+                lots = [lots];
+            }
+            // We don't constrain lot selection to already sold products
+            // as pos terminal doing inventory records only on session closing
```
Comment (cvinh) [inline pos_lot_selection/static/src/js/db.js]: As you noticed here https://github.com/OCA/pos/pull/827#discussion_r948780559 would it be better to filter lots to have only available in the linked warehouse ?

File: pos_lot_selection/i18n/es.po [hunk 6]
```diff
@@ -1,27 +0,0 @@
...
-"MIME-Version: 1.0\n"
-"Content-Type: text/plain; charset=UTF-8\n"
-"Content-Transfer-Encoding: \n"
-"Plural-Forms: \n"
-
-#. module: pos_lot_selection
-#. openerp-web
-#: code:addons/pos_lot_selection/static/src/xml/pos.xml:12
-#, python-format
-msgid "Select a Serial/Lot Number"
```
Comment (simahawk) [inline pos_lot_selection/i18n/es.po]: this should stay no?

File: pos_lot_selection/static/src/js/db.js (hunk 5 above)
Comment (simahawk) [reply pos_lot_selection/static/src/js/db.js]: is done in models.js

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js (hunk 1 above)
Comment (simahawk) [reply pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: Indeed. Moreover, now only the good ones are loaded, LG.

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            if (this.props.product) {
+                this.lots = this.env.pos.db.lot_by_product_id[this.props.product.id];
+            }
+        }
+        getPayload() {
+            return {
+                newArray: [Object.assign({}, {text: $("#new_select").val()})],
```
Comment (simahawk) [inline pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: I don't get why you should use Object.assign here...

File: pos_lot_selection/static/src/js/popup/ProductScreen.js (hunk 3 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/js/popup/ProductScreen.js]: I don't know if that is possible when the product and tracked and weighed, usually when it is tracked it is not divisible 
if so this need to be tasted with the electronic weights

File: pos_lot_selection/static/src/js/popup/LotSelectorPopup.js (hunk 1 above)
Comment (grindtildeath) [reply pos_lot_selection/static/src/js/popup/LotSelectorPopup.js]: Even with POS being designed to work offline, it can still work offline without having to load the lots. It's just that you'd have to enter the number manually instead of having the possibility to select, not having a lot loaded is not blocking in any way, as if you wouldn't have a product loaded and would be unable to sell it.

The possibility to select is therefore nice to have but again it's not blocking if you are offline, whereas loading the lots in the POS means that we have to sync the lots if there are many movements, to make sure they are up to date, what could easily be avoided by not loading the lots in the POS.

Finally, please have a look at pos_coupon, and you'll notice that the POS is doing a RPC request when a coupon code is entered on the point of sale, and it's impossible to apply a coupon in offline mode. I guess the reason is more or less the same than here, where you prefer to make sure that you're online to validate a coupon that having to load and sync all the time to make sure it's valid...

File: pos_lot_selection/static/src/js/popup/ProductScreen.js (hunk 3 above)
Comment (grindtildeath) [reply pos_lot_selection/static/src/js/popup/ProductScreen.js]: It is not divisible in case of serial numbers, but there could be lots of cases where a good tracked by lot has to be weighed. Think of food or pharma industries...

**Comment by grindtildeath** [comment]
@i-vyshnevska please check my last comoments as it's definitely blocking

**Comment by i-vyshnevska** [comment]
@grindtildeath updated

File: pos_lot_selection/models/stock_production_lot.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from odoo import api, fields, models
+from odoo.tools import float_compare
+
+
+class ProductionLot(models.Model):
+    _inherit = "stock.production.lot"
+
+    available_on_pos = fields.Boolean(compute="_compute_available_on_pos", store=True)
```
Comment (grindtildeath) [inline pos_lot_selection/models/stock_production_lot.py]: I guess we could get rid of this class by modifying the search from the POS to use the `product_qty` field? cf suggestion below.

File: pos_lot_selection/static/src/js/popup/ProductScreen.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                var lots = await this.rpc(
+                    {
+                        model: "stock.production.lot",
+                        method: "search_read",
+                        kwargs: {
+                            domain: [
+                                "&",
+                                "&",
+                                ["product_id", "=", product.id],
+                                ["available_on_pos", "=", true],
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/ProductScreen.js]:
```suggestion
                                ["product_qty", ">", 0],
```

File: pos_lot_selection/static/src/js/popup/OrderWidget.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    const Registries = require('point_of_sale.Registries');
+    const OrderWidget = require('point_of_sale.OrderWidget');
+
+    
+    var CustomOrderWidget = (OrderWidget) => class extends OrderWidget{
+        async _editPackLotLines(event) {
+            const orderline = event.detail.orderline;
+            
+            this.env.session.lots = await this.get_lots(orderline.product, this.env.session);
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/OrderWidget.js]: I'm not sure if the session object is the right place where to store the possible values. Maybe the POS DB? @ivantodorovich any better idea?

File: pos_lot_selection/static/src/js/popup/OrderWidget.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            return super._editPackLotLines(event);
+        }
+
+        async get_lots(product){
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/OrderWidget.js]: I guess the best place where to define this function is in the POS DB, so that you could call it using `this.env.pos.db.get_lots` in both cases.

File: pos_lot_selection/static/src/js/popup/ProductScreen.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    const PosLotSaleProductScreen = (ProductScreen) =>
+        class extends ProductScreen {
+            async _getAddProductOptions(product, base_code) {
+                if (["serial", "lot"].includes(product.tracking)) {
+                    // Set lots to context as it one possible way to extract arguments
+                    // later in popup
+                    this.env.session.lots = await this.get_lots(product);
+                } else {
+                    // Cleanup
+                    this.env.session.product = false;
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/ProductScreen.js]: Why is this needed? :thinking:

File: pos_lot_selection/static/src/js/popup/ProductScreen.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+                }
+
+                return super._getAddProductOptions(product, (base_code = base_code));
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/ProductScreen.js]: Shouldn't we reset the value of lots in the session (or in the DB cf comment above) ?
```suggestion
                res = super._getAddProductOptions(product, (base_code = base_code));
                this.env.session.lots = []
                return res
```

File: pos_lot_selection/static/src/js/popup/EditListPopup.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    "use strict";
+
+    const EditListPopup = require("point_of_sale.EditListPopup");
+    const Registries = require("point_of_sale.Registries");
+
+    const LotSelectEditListPopup = (EditListPopup) =>
+        class extends EditListPopup {
+            constructor() {
+                super(...arguments);
+                this.props.lots = this.env.session.lots;
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/EditListPopup.js]: I guess we could do this to make sure the value is set only if we are on the `EditListPopup` for lot selection. (in case this Popup is used in another module)
```suggestion
                if (this.props.title === this.env._t("Lot/Serial Number(s) Required")) {
                    this.props.lots = this.env.session.lots;
                }
```

File: pos_lot_selection/static/src/xml/LotSelectorPopup.xml [hunk 15]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2022 Camptocamp SA
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl). -->
+<templates id="template" xml:space="preserve">
+
+    <t t-inherit="point_of_sale.EditListPopup" t-inherit-mode="extension">
+
+        <xpath expr="//t[@t-foreach='state.array']" position="replace">
+            <t t-foreach="state.array" t-as="item" t-key="item._id">
+                <EditListInput item="item" lots="props.lots" />
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/xml/LotSelectorPopup.xml]: Shouldn't we do this only if `props.lots` is defined? To avoid messing with possible other usages of `EditListPopup`

File: pos_lot_selection/static/src/xml/LotSelectorPopup.xml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            </t>
+        </xpath>
+    </t>
+
+
+    <t t-inherit="point_of_sale.EditListInput" t-inherit-mode="extension">
+
+        <xpath expr="//input" position="attributes">
+            <attribute name="list">prepared_lots</attribute>
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/xml/LotSelectorPopup.xml:24]: Same as above, maybe to all of this only if we have `props.lots` defined?

**Review by grindtildeath — COMMENTED** [review]
Looks much better overall, thank you :+1: 

IMO there are still a few points to be improved, @ivantodorovich can you have a look please?

File: pos_lot_selection/static/src/js/popup/OrderWidget.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    const Registries = require('point_of_sale.Registries');
+    const OrderWidget = require('point_of_sale.OrderWidget');
+
+    
+    var CustomOrderWidget = (OrderWidget) => class extends OrderWidget{
+        async _editPackLotLines(event) {
+            const orderline = event.detail.orderline;
+            
+            this.env.session.lots = await this.get_lots(orderline.product, this.env.session);
+            return super._editPackLotLines(event);
```
Comment (grindtildeath) [inline pos_lot_selection/static/src/js/popup/OrderWidget.js]: Forgot about this, but maybe we should "reset" the values of lots that are stored after the call to super (as I wrote for the call in `ProductScreen`)

File: pos_lot_selection/static/src/js/popup/ProductScreen.js (hunk 9 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/js/popup/ProductScreen.js]: product_qty is not a stored field so we need to use a filter here, having a boolean field looks like more explicit way

File: pos_lot_selection/__manifest__.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    "version": "15.0.1.0.0",
+    "category": "Point of Sale",
+    "author": "Tecnativa, Camptocamp, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/pos",
+    "license": "AGPL-3",
+    "depends": [
+        "point_of_sale",
+    ],
+    "assets": {
+        "web.assets_qweb": ["pos_lot_selection/static/src/**/*"],
```
Comment (ivantodorovich) [inline pos_lot_selection/__manifest__.py]:
```suggestion
        "web.assets_qweb": ["pos_lot_selection/static/src/xml/**/*.xml"],
```

We don't want to load everything here, just the xml files

File: pos_lot_selection/models/stock_production_lot.py (hunk 8 above)
Comment (ivantodorovich) [reply pos_lot_selection/models/stock_production_lot.py]: I agree this is not cool.

As it's a stored computed field, it will recompute everytime the quants quantities change, which is on every stock.move for this product.. it's totally unnecessary.

It's even worse when we consider the quants in the customer location, where all delivered lots will be linked to in the end. It won't scale well as the db grows.

I think you could improve this by making it NOT stored, and implement a search method instead.

Alternatively, and maybe even better.. consider exposing the list of lots in a public method instead, specifically built for the pos. This means, in js, you won't call `search_read` directly.. you'd call this method instead.

For example, it could be a method in the `product.product` model: `get_available_lots_for_pos`. This allows you to do some python processing/filtering before sending back the records.

File: pos_lot_selection/models/stock_production_lot.py [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+
+    @api.depends("quant_ids.quantity")
+    def _compute_available_on_pos(self):
+        for lot in self:
+            lot.available_on_pos = (
+                float_compare(lot.product_qty, 0, precision_digits=2) > 0
```
Comment (ivantodorovich) [inline pos_lot_selection/models/stock_production_lot.py]: FTR, it you are keeping this, `precision_digits` shouldn't be hardcoded.
It should be read from the product's uom

File: pos_lot_selection/static/src/js/popup/OrderWidget.js (hunk 10 above)
Comment (ivantodorovich) [reply pos_lot_selection/static/src/js/popup/OrderWidget.js]: I don't think we need to store them at all.
IMO the lots should be fetched from the server in the EditListPopup popup, before it's shown, and that's it

File: pos_lot_selection/static/src/js/popup/OrderWidget.js (hunk 11 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/js/popup/OrderWidget.js]: appeared db implementation doesn't have rpc component so I used PosModel instead

File: pos_lot_selection/static/src/js/popup/ProductScreen.js (hunk 13 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/js/popup/ProductScreen.js]: this will not work as it is async method

File: pos_lot_selection/static/src/js/popup/OrderWidget.js (hunk 10 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/js/popup/OrderWidget.js]: EditListPopup is synchronous so to be able to fetch products there we need to load lots on session start, that implementation was declined 
second, is that there is no product in EditListPopup, so I need somehow to send the product there, or override part of the method, but that brings another issue as default implementation can call a chain of popups to configure the product and get a weight and we have an issue because lot popup is in the middle of that chain

File: pos_lot_selection/static/src/xml/LotSelectorPopup.xml (hunk 16 above)
Comment (i-vyshnevska) [reply pos_lot_selection/static/src/xml/LotSelectorPopup.xml:24]: that is handled fine datalist will be empty

**Review by grindtildeath — APPROVED** [review]
To me looks much better right now, thank you :+1: 

Might still not be perfect but enough to be usable, so you get my approval.

**Review by ivantodorovich — APPROVED** [review]
